extensions/amp-analytics/0.1/visibility-impl.js

```diff
@@ -370,22 +371,43 @@ export class Visibility {
 
     const visible = change.intersectionRatio * 100;
     for (let c = listeners.length - 1; c >= 0; c--) {
-      const shouldBeVisible = !!listeners[c]['shouldBeVisible'];
-      if (this.updateCounters_(visible, listeners[c], shouldBeVisible)
-          && this.viewer_.isVisible() == shouldBeVisible) {
-        this.prepareStateForCallback_(
-            listeners[c]['state'], change.boundingClientRect);
-        listeners[c].callback(listeners[c]['state']);
+      const shouldBeVisible = !!listeners[c].shouldBeVisible;
+      const config = listeners[c].config;
+      const state = listeners[c].state;
+      if (state[SCHEDULED_RUN_ID]) {
+        this.timer_.cancel(state[SCHEDULED_RUN_ID]);
+        state[SCHEDULED_RUN_ID] = null;
+      }
+      // Update states and check if all conditions are satisfied
+      if (this.updateCounters_(visible, listeners[c], shouldBeVisible)) {
+        this.prepareStateForCallback_(state, change.boundingClientRect);
+        listeners[c].callback(state);
         listeners.splice(c, 1);
+      } else if (state[IN_VIEWPORT]) {
+        // There is unmet duration condition, schedule a check
+        const timeToWait = this.computeTimeToWait_(state, config);
+        if (timeToWait <= 0) {
+          continue;
+        }
+        state[SCHEDULED_RUN_ID] = this.timer_.delay(() => {
```

Comment:
The debounce function does seem to be useful. I prefer to introduce in a separate PR, probably as a reusable util.

For the function instance thing, please note that the function takes parameters. Not sure how we can use one instance and pass in different params. Can you give an example.